sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -151,6 +164,30 @@ abstract class Expression extends TreeNode[Expression] {
   }
 
   /**
+   * Returns the hash for this expression. Expressions that compute the same result, even if
+   * they differ cosmetically should return the same hash.
+   */
+  def semanticHash() : Int = {
```

Comment:
When you look at the methods `equals`, `semanticEquals` and `hashCode` of the `AttributeReference`, you will see that they are not matched, as `equals` will take consideration of the `name`, but the other 2 are not, that's why I am thinking we can also use the `hashCode`, instead of adding the new method `semanticHash`.

Anyway, it's not an external API, we can change it back anytime, as 1.6 is almost code freeze, and this is critical for people now.